testutil/integration/factory/factory.go

```diff
@@ -0,0 +1,326 @@
+	// DoEthTx builds, signs and broadcast an Ethereum tx with the provided private key and txArgs
+	DoEthTx(privKey cryptotypes.PrivKey, txArgs evmtypes.EvmTxArgs) (abcitypes.ResponseDeliverTx, error)
+	// DoCosmosTx builds, signs and broadcast a Cosmos tx with the provided private key and txArgs
+	DoCosmosTx(privKey cryptotypes.PrivKey, txArgs CosmosTxArgs) (abcitypes.ResponseDeliverTx, error)
+	// EstimateGasLimit estimates the gas limit for a tx with the provided private key and txArgs
+	EstimateGasLimit(from *common.Address, txArgs *evmtypes.EvmTxArgs) (uint64, error)
+}
+
+var _ TxFactory = (*IntegrationTxFactory)(nil)
+
+// IntegrationTxFactory is a helper struct to build and broadcast transactions
+// to the network on integration tests. This is to simulate the behavior of a real user.
+type IntegrationTxFactory struct {
+	grpcHandler grpc.Handler
+	network     network.Network
+}
+
+// DeployContract deploys a contract with the provided private key,
+// compiled contract data and constructor arguments
+func (tf *IntegrationTxFactory) DeployContract(
+	priv cryptotypes.PrivKey,
+	contract evmtypes.CompiledContract,
+	constructorArgs ...interface{},
+) (common.Address, error) {
+	// Get account's nonce to create contract hash
+	from := common.BytesToAddress(priv.PubKey().Address().Bytes())
+	account, err := tf.grpcHandler.GetEvmAccount(from)
+	if err != nil {
+		return common.Address{}, err
+	}
+	nonce := account.GetNonce()
+
+	ctorArgs, err := contract.ABI.Pack("", constructorArgs...)
+	if err != nil {
+		return common.Address{}, err
+	}
+	data := contract.Bin
+	data = append(data, ctorArgs...)
+
+	args := evmtypes.EvmTxArgs{
+		Input: data,
+		Nonce: nonce,
+	}
+	res, err := tf.DoEthTx(priv, args)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if err := checkEthTxResponse(&res); err != nil {
+		return common.Address{}, err
+	}
+	return crypto.CreateAddress(from, nonce), nil
+}
+
+// DoContractCall executes a contract call with the provided private key and txArgs
+// It first builds a MsgEthereumTx and then broadcast it to the network.
+func (tf *IntegrationTxFactory) DoEthTx(
+	priv cryptotypes.PrivKey,
+	txArgs evmtypes.EvmTxArgs,
+) (abcitypes.ResponseDeliverTx, error) {
+	msgEthereumTx, err := tf.createMsgEthereumTx(priv, txArgs)
+	if err != nil {
+		return abcitypes.ResponseDeliverTx{}, err
+	}
+
+	signedMsg, err := signMsgEthereumTx(msgEthereumTx, priv, tf.network.GetChainID())
+	if err != nil {
+		return abcitypes.ResponseDeliverTx{}, err
+	}
+
+	txBytes, err := buildAndEncodeEthTx(signedMsg, tf.network.GetDenom())
+	if err != nil {
+		return abcitypes.ResponseDeliverTx{}, err
+	}
+
+	_, err = tf.network.Simulate(txBytes)
+	if err != nil {
+		return abcitypes.ResponseDeliverTx{}, err
+	}
+
+	res, err := tf.network.BroadcastTxSync(txBytes)
+	if err != nil {
+		return abcitypes.ResponseDeliverTx{}, err
+	}
+
+	if err := checkEthTxResponse(&res); err != nil {
```

Comment:
the latter

Reply:
I think this one is a key point, its what allowed me to detect the issues on the revenue module. This will return the error if the transaction actually failed, if this is expected to happen then that error should be handled in the test that calls this function explicitely instead of returning a false succesfull transaction wdyt?